### PR TITLE
feat: support deterministic deployment proxy contract transaction (#2287)

### LINKED ISSUE #2287: Foundry Create2 Support
### Problem

Foundry `CREATE2` requires a predeployed deterministic deployment proxy to be available at `0x4e59b44847b379578588920ca78fbf26c0b4956c`.

The proxy's repo is here: https://github.com/Arachnid/deterministic-deployment-proxy
The way to deploy it is with a Pre EIP 155 transaction, which are now supported on Hedera.
Trying to send the transaction that creates it results in an error of low gas price
```shell
cast publish --rpc-url  https://testnet.hashio.io/api 0xf8a58085174876e800830186a08080b853604580600e600039806000f350fe7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe03601600081602082378035828234f58015156039578182fd5b8082525050506014600cf31ba02222222222222222222222222222222222222222222222222222222222222222a02222222222222222222222222222222222222222222222222222222222222222
```

Results in an error:
```
Error: 
(code: -32009, message: [Request ID: fbda5b21-041b-4e61-b4bf-c6a814304aac] Gas price '100000000000' is below configured minimum gas price '830000000000', data: None)
```

Increasing the Gas price would result in the address of the proxy changing, so it's not possible.

### Solution

Find a workaround to deploy the proxy to Hedera, to increase compability with Foundry and EVM equivalence.

### Alternatives

-

### PR DESCRIPTION
**Pull request by quiet-node**
**Description**:
This PR adds support for the Foundry deterministic deployment proxy contract transaction, and add acceptance tests to cover the new support.

**Related issue(s)**:

Fixes #2287 

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: packages/relay/src/lib/constants.ts [hunk 1]
```diff
@@ -175,4 +175,10 @@ export default {
     NEW_HEADS: 'newHeads',
     NEW_PENDING_TRANSACTIONS: 'newPendingTransactions',
   },
+
+  // @source: Foundry related constants below can be found at https://github.com/Arachnid/deterministic-deployment-proxy?tab=readme-ov-file#latest-outputs
+  FOUNDRY_DETERMINISTIC_DEPLOYER_TRANSACTION:
```
Comment (Nana-EC) [inline packages/relay/src/lib/constants.ts:180]: nit: Naming should be more general
```suggestion
  DETERMINISTIC_DEPLOYER_PROXY_TRANSACTION:
```

File: packages/relay/src/lib/precheck.ts [hunk 2]
```diff
@@ -152,6 +158,15 @@ export class Precheck {
     }
   }
 
+  /**
+   * Checks if a transaction is the Foundry deterministic deployment transaction.
+   * @param {Transaction} tx - The transaction to check.
+   * @returns {boolean} Returns true if the transaction is the Foundry deterministic deployment transaction, otherwise false.
+   */
+  static isFoundryDeterministicDeploymentTransaction(tx: Transaction): boolean {
```
Comment (Nana-EC) [inline packages/relay/src/lib/precheck.ts:166]: Add coverage in the precheck.spec.ts
Always add unit tests first. Then component/integration level tests, then acceptance/E2E tests

**Review by Nana-EC — COMMENTED** [review]
Logic is good.
Some naming and test suggestions

File: packages/relay/src/lib/constants.ts (hunk 1 above)
Comment (quiet-node) [reply packages/relay/src/lib/constants.ts]: Copy!

File: packages/relay/src/lib/precheck.ts (hunk 2 above)
Comment (quiet-node) [reply packages/relay/src/lib/precheck.ts]: Copy!

File: packages/server/tests/acceptance/rpc_batch1.spec.ts [hunk 3]
```diff
@@ -817,6 +817,37 @@ describe('@api-batch-1 RPC Server Acceptance Tests', function () {
...
+        const receipt = await mirrorNode.get(
+          `/contracts/results/${foundryDeterministicDeployTransactionHash}`,
+          requestId,
+        );
+        const fromAccountInfo = await global.mirrorNode.get(`/accounts/${receipt.from}`);
+
+        expect(receipt).to.exist;
+        expect(fromAccountInfo.evm_address).to.eq(constants.DETERMINISTIC_DEPLOYMENT_SIGNER);
+        // notice: the assertion below is currently blocked by the Services
+        // expect(receipt.address).to.eq(constants.DETERMINISTIC_PROXY_CONTRACT);
```
Comment (AlfredoG87) [inline packages/server/tests/acceptance/rpc_batch1.spec.ts:848]: should we open a ticket so we don't forget to uncomment this line?

File: packages/ws-server/tests/acceptance/sendRawTransaction.spec.ts [hunk 4]
```diff
@@ -96,6 +105,31 @@ describe('@release @web-socket eth_sendRawTransaction', async function () {
...
+      );
+
+      const receipt = await global.mirrorNode.get(
+        `/contracts/results/${foundryDeterministicDeploymentTransactionHash.result}`,
+      );
+      const fromAccountInfo = await global.mirrorNode.get(`/accounts/${receipt.from}`);
+      expect(receipt).to.exist;
+      expect(fromAccountInfo.evm_address).to.eq(constants.DETERMINISTIC_DEPLOYMENT_SIGNER);
+      // notice: the assertion below is currently blocked by the Services
+      // expect(receipt.address).to.eq(constants.DETERMINISTIC_PROXY_CONTRACT);
```
Comment (AlfredoG87) [inline packages/ws-server/tests/acceptance/sendRawTransaction.spec.ts:131]: same as above. in same ticket.

**Review by AlfredoG87 — APPROVED** [review]
LGTM. 

Just 1 suggestion to open a follow-up ticket due to the commented line on the assertions.

File: packages/server/tests/acceptance/rpc_batch1.spec.ts (hunk 3 above)
Comment (quiet-node) [reply packages/server/tests/acceptance/rpc_batch1.spec.ts:848]: Ah I was gonna wait until the fix in the services to be included in the next releases which I believe will happen soon, then I'll uncomment this and we can merge it to main. Without the fix in the services, this feature would not work at all. I'm setting this PR as draft for early review and to test out the CIs.

**Comment by Nana-EC** [comment]
@quiet-node you should be able to test this with `v0.49.0` of services now

File: packages/server/tests/acceptance/rpc_batch1.spec.ts (hunk 3 above)
Comment (Nana-EC) [reply packages/server/tests/acceptance/rpc_batch1.spec.ts:848]: Fix is available in `v0.49.0` now. Might need a local node update

**Comment by quiet-node** [comment]
> @quiet-node you should be able to test this with `v0.49.0` of services now

Thanks for the heads up! Will do!

**Review by Nana-EC — APPROVED** [review]

**Comment by quiet-node** [comment]
https://github.com/hashgraph/hedera-json-rpc-relay/pull/2393

**Pull request closed without merging**